Support Dynamic Complex Types

MarkusHorstmann opened the pull request:
## Proposed changes

Complex Type support assumes that every OPC type is represented by exactly one .Net type, which in turn knows encoding ids, XML names etc.
These changes enable use of a single type that implements encoding/decoding based on OPC type information at runtime.

## Related Issues

- Fixes #2145 

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply. You can also fill these out after creating the PR._

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] Enhancement (non-breaking change which adds functionality)
- [ ] Test enhancement (non-breaking change to increase test coverage)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected, requires version increase of Nuget packages)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/OPCFoundation/UA-.NETStandard/blob/master/CONTRIBUTING.md) doc.
- [x] I have signed the [CLA](https://opcfoundation.org/license/cla/ContributorLicenseAgreementv1.0.pdf).
- [x] I ran tests locally with my changes, all passed.
- [ ] I fixed all failing tests in the CI pipelines. 
- [ ] I fixed all introduced issues with CodeQL and LGTM.
- [ ] I have added tests that prove my fix is effective or that my feature works and increased code coverage.
- [ ] I have added necessary documentation (if appropriate).
- [ ] Any dependent changes have been merged and published in downstream modules.

mregen commented:
Hi @MarkusHorstmann , please elaborate how this extension can be used, or even better provide a test case / sample. I already cherry picked your bug fix, thanks!

mregen commented on Stack/Opc.Ua.Core/Types/Encoders/IComplexTypeInstance.cs:
```diff
@@ -25,4 +27,29 @@ public interface IComplexTypeInstance
+        /// Returns the NodeId for the default XML encoding for the type.
+        /// </summary>
+        /// <value>The NodeId for the  XML encoding id.</value>
+        ExpandedNodeId XmlEncodingId { get; set; }
```
is there no JSON encoding ID needed?

MarkusHorstmann replied:
Good catch. Actually neither of these are needed - they are already on IEncodeable. I've removed them.

MarkusHorstmann commented:
> Hi @MarkusHorstmann , please elaborate how this extension can be used, or even better provide a test case / sample. I already cherry picked your bug fix, thanks!

I have added test cases using a dynamic encodeable with simple type information (just a list of string-valued fields). For a realistic use case feel free to look at the [CESMII NodeSet utilities](https://github.com/cesmii/CESMII-NodeSet-Utilities/blob/main/src/CESMII.OpcUa.NodeSetModel.Factory.Opc/DynamicComplexType.cs). It is not meant to be a sample, though.

mregen commented:
Hi @MarkusHorstmann, looks pretty good, any idea why the tests are failing?

mregen commented:
[Check failure on line 1 in Opc.Ua.Core.Tests.Types.Encoders.JsonEncoderTests.Test_ExtensionObjectWithDynamicEncodeable](https://github.com/OPCFoundation/UA-.NETStandard/pull/2146/files#annotation_11734298883) 


```
@azure-pipelinesazure-pipelines
/ OPCFoundation.UA-.NETStandard
Opc.Ua.Core.Tests.Types.Encoders.JsonEncoderTests.Test_ExtensionObjectWithDynamicEncodeable
System.InvalidOperationException : Collection was modified; enumeration operation may not execute.
```

mregen reviewed (changes requested):
please fix the test cases

MarkusHorstmann commented:
> please fix the test cases

The new tests should now run fine. There are two other tests failing (ReadOnDiscoveryChannel and AppVeyor build) that don't seem to be related to my changes, and I've seen them fail in other branches as well. Please let me know if you think otherwise.

MarkusHorstmann commented:
Oops, thought I was closing the review comment, not the PR. Reopened it.

mregen commented on Stack/Opc.Ua.Core/Types/Encoders/XmlDecoder.cs:
```diff
@@ -2745,7 +2745,7 @@ private Array ReadArrayElements(string fieldName, BuiltInType builtInType, Type
             // skip whitespace.
-            while (m_reader.NodeType != XmlNodeType.Element)
+            while (m_reader.NodeType != XmlNodeType.Element && m_reader.NodeType != XmlNodeType.EndElement)
             {
```
Is this really needed? There is some concern this extra check can cause decoding issues.

MarkusHorstmann replied:
It tripped during testing but I can't reproduce anymore. I'll remove it and bring it back as a bug fix later if necessary.

gabrielgriesser commented:
Hello, I'm interested about this feature. I have to create Structure Type (so ExtensionObject) XML from my autoDiscovery code. 

I am able to auto discover my StructureDefinition :

`
foreach (var result in browseResults)
        {
            if (result.NodeClass != NodeClass.DataType)
                continue;

            var varNodeId = NodeId.Parse(result.NodeId.ToString());

            if (Session.ReadNode(varNodeId) is DataTypeNode dataTypeNode && (dataTypeNode.DataTypeDefinition != null))
                if (result.DisplayName.Text.Contains("ST_"))
                {
                    Debug.WriteLine("     DatTypeNode = {0}, NodeClass = {1}, NodeID = {2}, NodeFields = {3}", dataTypeNode.DisplayName.Text, dataTypeNode.NodeClass, dataTypeNode,
                        dataTypeNode.DataTypeDefinition);
                    var sd = dataTypeNode.DataTypeDefinition.Body as StructureDefinition;

                    // TODO Create my own XML from sd structure
                }


            BrowseRecursive(browser, varNodeId);
        }
`

But now, how can I dynamically create an XML file and how to use it to decode my ExtensionObject when I read my variable using `Session.ReadValue()` ?

MarkusHorstmann commented on Stack/Opc.Ua.Core/Types/Encoders/IComplexTypeInstance.cs:
```diff
@@ -25,4 +27,17 @@ public interface IComplexTypeInstance
+        /// </summary>
+        /// <returns>The xml qualified name for the type instance</returns>
+        XmlQualifiedName GetXmlName(IServiceMessageContext context);
+    }
```
A IComplexTypeInstance implementation needs the Encodable Factory and the Namespace Table from the context. The reason is that the implementation may need to make additional nested types in a structure available via the Factory. Otherwise, all known types would have to be pre-registered, which largely defeats the purpose of a truly dynamic (= meta-data-driven) type implementation.

MarkusHorstmann replied:
@mregen Please let me know if you need any additional information or maybe want to discuss on a call?

mregen replied:
Hi @MarkusHorstmann , we discussed the PR yesterday in the user group and it is accepted, just taking a final check then its ready to go. Thanks!

MarkusHorstmann commented on Stack/Opc.Ua.Core/Types/Encoders/XmlEncoder.cs:
```diff
@@ -1933,10 +1933,12 @@ public void WriteExtensionObjectBody(object body)
-            XmlQualifiedName xmlName = EncodeableFactory.GetXmlName(encodeable.GetType());
+            XmlQualifiedName xmlName = EncodeableFactory.GetXmlName(encodeable, this.Context);
+            PushNamespace(xmlName.Namespace);
             m_writer.WriteStartElement(xmlName.Name, xmlName.Namespace);
```
Yes, makes sense. Removed.

MarkusHorstmann commented:
> Hello, I'm interested about this feature. I have to create Structure Type (so ExtensionObject) XML from my autoDiscovery code.
> 
> I am able to auto discover my StructureDefinition :
> 
> ` foreach (var result in browseResults) { if (result.NodeClass != NodeClass.DataType) continue;
> 
> ```
>         var varNodeId = NodeId.Parse(result.NodeId.ToString());
> 
>         if (Session.ReadNode(varNodeId) is DataTypeNode dataTypeNode && (dataTypeNode.DataTypeDefinition != null))
>             if (result.DisplayName.Text.Contains("ST_"))
>             {
>                 Debug.WriteLine("     DatTypeNode = {0}, NodeClass = {1}, NodeID = {2}, NodeFields = {3}", dataTypeNode.DisplayName.Text, dataTypeNode.NodeClass, dataTypeNode,
>                     dataTypeNode.DataTypeDefinition);
>                 var sd = dataTypeNode.DataTypeDefinition.Body as StructureDefinition;
> 
>                 // TODO Create my own XML from sd structure
>             }
> 
> 
>         BrowseRecursive(browser, varNodeId);
>     }
> ```
> 
> `
> 
> But now, how can I dynamically create an XML file and how to use it to decode my ExtensionObject when I read my variable using `Session.ReadValue()` ?

Hi, Gabriel! This Pull Request only provides the necessary hooks in the Core to make it possible implement dynamic types. I have implemented a dynamic type that I need in the context of a NodeSet editing tool, and it uses the editing tool's type system not the OPC SDK's type definition classes. The code is publically available here (https://github.com/cesmii/CESMII-NodeSet-Utilities/blob/main/src/CESMII.OpcUa.NodeSetModel.Factory.Opc/DynamicComplexType.cs). It should be relatively straightforward to change it to read the type information from a DataTypeDefinition instead of the NodeModel's DataTypeModel.

I also implemented a [test case](https://github.com/MarkusHorstmann/UA-.NETStandard/blob/1ad98297952964015ab4a0bcc363bee07e304fab/Tests/Opc.Ua.Core.Tests/Types/Encoders/EncoderCommon.cs#L822) that implements a rudimentary dynamic type that handles only string values. This could be another starting point.

mregen commented:
@MarkusHorstmann please merge master into your PR, the ci VMs had been updated recently so the failing build will not pass otherwise and i can't merge.
